crates/turborepo-lib/src/engine/mod.rs

```diff
@@ -110,4 +114,110 @@ impl Engine<Built> {
                 .collect(),
         )
     }
+
+    pub fn validate(
+        &self,
+        package_graph: &PackageGraph,
+        concurrency: u32,
+    ) -> Result<(), Vec<ValidateError>> {
+        // TODO(olszewski) once this is hooked up to a real run, we should
+        // see if using rayon to parallelize would provide a speedup
+        let (persistent_count, mut validation_errors) = self
+            .task_graph
+            .node_indices()
+            .map(|node_index| {
+                let TaskNode::Task(task_id) = self
+                    .task_graph
+                    .node_weight(node_index)
+                    .expect("graph should contain weight for node index")
+                else {
+                    // No need to check the root node if that's where we are.
+                    return Ok(false);
+                };
+                let is_persistent = self
+                    .task_definitions
+                    .get(task_id)
+                    .map_or(false, |task_def| task_def.persistent);
+
+                for dep_index in self
+                    .task_graph
+                    .neighbors_directed(node_index, petgraph::Direction::Outgoing)
+                {
+                    let TaskNode::Task(dep_id) = self
+                        .task_graph
+                        .node_weight(dep_index)
+                        .expect("node index not in graph")
```

Comment:
I would like to see a messeage like 'dep_index comes from iterating the graph and must be there'

Seems like a very common thing to need, I'm surprised they don't have an access pattern that simplifies this...

Reply:
I agree that it's surprising that there isn't an "unwrap-less" method for getting neighbor weights. There's [`IntoNodeReferences`](https://docs.rs/petgraph/latest/petgraph/visit/trait.IntoNodeReferences.html) which bundles the id with the ability to fetch the weight, but it isn't implemented for any of the neighbor iterators. Maybe on a DYF I'll go through and try to implement it for some more iterators.